### PR TITLE
Support bitwise allreduce in NCCL communicator

### PR DESCRIPTION
**Pull request by rongou**
Use `AllGather` to collect all the data first, then do the bitwise allreduce locally.

### CODE REVIEW

**Comment by rongou** [comment]
@trivialfis

**Review by trivialfis — APPROVED** [review]

**Comment by trivialfis** [comment]
Apologies for the oversight, the no_sync is a CTK-12 feature.

**Comment by rongou** [comment]
Switched to use `LaunchN`.

File: src/collective/nccl_device_communicator.cu [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+  auto n_uniques = std::distance(converted.begin(), iter);
+
+  CHECK_EQ(n_uniques, world)
+      << "Multiple processes within communication group running on same CUDA "
+      << "device is not supported. " << PrintUUID(s_this_uuid) << "\n";
+
+  nccl_unique_id_ = GetUniqueId();
+  dh::safe_cuda(cudaSetDevice(device_ordinal_));
+  dh::safe_nccl(ncclCommInitRank(&nccl_comm_, world, nccl_unique_id_, rank));
+  dh::safe_cuda(cudaStreamCreate(&cuda_stream_));
```
Comment (trivialfis) [inline src/collective/nccl_device_communicator.cu:51]: Just a side note, there's `dh::CUDAStream`, which is a RAII version of async stream.

Comment (rongou) [reply src/collective/nccl_device_communicator.cu:51]: Switched to it.

**Pull request merged**